### PR TITLE
Add back button to wizard

### PR DESCRIPTION
**Pull request by EricJizbaMSFT**
I took another look at using the more advanced multi-step UI (added to VS Code api almost half a year ago now 😬) and kind of went ham implementing it this weekend lol

Fixes https://github.com/Microsoft/vscode-azuretools/issues/220. Based on [this sample extension](https://github.com/Microsoft/vscode-extension-samples/tree/master/quickinput-sample)

Functions demo:
backBtn1

Cosmos demo:
backBtn2

Ran into quite a few edge cases so added quite a few unit tests.

### CODE REVIEW

File: ui/index.d.ts [hunk 1]
```diff
@@ -553,27 +565,63 @@ export interface IAzureQuickPickOptions extends QuickPickOptions {
...
+     */
+    promptSteps: AzureWizardPromptStep<T>[];
+
+    /**
+     * The steps to execute, in order
+     */
+    executeSteps?: AzureWizardExecuteStep<T>[];
+}
+
+export interface IWizardOptions<T> extends ISubWizardOptions<T> {
```
Comment (nturinski) [inline ui/index.d.ts]: So weird to me that the main class extends the sub class lol

File: ui/src/wizard/AzureWizard.ts [hunk 2]
```diff
@@ -3,57 +3,130 @@
...
-    private readonly _subWizards: AzureWizard<T>[] = [];
-    private _wizardContext: T;
+    private readonly _finishedPromptSteps: AzureWizardPromptStep<T>[] = [];
+    private readonly _wizardContext: T;
 
-    public constructor(promptSteps: AzureWizardPromptStep<T>[], executeSteps: AzureWizardExecuteStep<T>[], wizardContext: T) {
-        this._promptSteps = promptSteps;
-        this._executeSteps = executeSteps;
+    public constructor(wizardContext: T, options: types.IWizardOptions<T>) {
+        // reverse steps to make it easier to use push/pop
```
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply ui/src/wizard/AzureWizard.ts]: I agree reversing the array is confusing which is why I wrote unit tests to specifically cover the order of the steps 😊 (for example [this one](https://github.com/Microsoft/vscode-azuretools/pull/444/files#diff-d7e5b2a59ca8eb655c6130c88e34b248R220)). In terms of your suggestion, I think `shift` and `unshift` is also confusing (I'm much more familiar with push/pop). As a tie-breaker between two confusing options - I think push/pop is more efficient

File: ui/src/wizard/AzureWizard.ts [hunk 3]
```diff
@@ -3,57 +3,130 @@
...
+    public get totalSteps(): number {
+        return this._finishedPromptSteps.length + this._promptSteps.filter(s => s.shouldPrompt(this._wizardContext)).length + 1;
+    }
+
+    public async prompt(actionContext: types.IActionContext): Promise<void> {
+        // Insert Wizard UI into ext.ui.rootUserInput - to be used instead of vscode.window UI
+        let oldRootUserInput: IRootUserInput = vscode.window;
+        if (ext.ui instanceof AzureUserInput) {
+            oldRootUserInput = ext.ui.rootUserInput;
+            ext.ui.rootUserInput = new AzureWizardUserInput(this);
```
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply ui/src/wizard/AzureWizard.ts]: If `new AzureWizardUserInput(this);` throws an error, then `ext.ui.rootUserInput` never gets assigned and we don't need to execute the finally block

File: ui/src/wizard/AzureWizardUserInput.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                        // Only back button is supported for now
+                        reject(new GoBackError());
+                    }),
+                    quickPick.onDidHide(() => {
+                        reject(new UserCancelledError());
+                    })
+                );
+
+                // Show progress bar while loading quick picks
+                quickPick.busy = true;
```
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply ui/src/wizard/AzureWizardUserInput.ts]: That's already possible with the standard showQuickPick. You just have to pass in `Promise<QuickPick[]>` instead of `QuickPick[]`

File: ui/src/wizard/StorageAccountListStep.ts [hunk 5]
```diff
@@ -61,53 +60,52 @@ export enum StorageAccountReplication {
     RAGRS = 'RAGRS'
```
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply ui/src/wizard/StorageAccountListStep.ts]: If you expand the diff, there's a comment saying what it stands for right above this

File: ui/index.d.ts (hunk 1 above)
Comment (EricJizbaMSFT) [reply ui/index.d.ts]: I see what you mean - it's weird because I'm talking about the sub-wizard, not the sub-class. Although these are both interfaces, not classes lol. Anyways let me know if you think of a better name

File: ui/src/wizard/StorageAccountCreateStep.ts [hunk 6]
```diff
@@ -5,44 +5,44 @@
...
+                location: newLocation
+            }
+        );
+        const createdStorageAccount: string = localize('CreatedStorageAccount', 'Successfully created storage account "{0}".', newName);
+        ext.outputChannel.appendLine(createdStorageAccount);
+    }
 
-        return wizardContext;
+    public shouldExecute(wizardContext: T): boolean {
+        return !wizardContext.storageAccount;
```
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply ui/src/wizard/StorageAccountCreateStep.ts]: Here is the difference between these two:
1. `newStorageAccountName`: If this is not defined, this should throw an error
1. `storageAccount`: If this is defined, we can silently skip this step
